framework/src/modules/dpos_v2/commands/vote_delegate.ts

```diff
@@ -157,17 +189,39 @@ export class VoteDelegateCommand extends BaseCommand {
 			if (vote.amount < BigInt(0)) {
 				// unvote
 				if (originalUpvoteIndex < 0) {
-					throw new Error('Cannot cast downvote to delegate who is not upvoted.');
+					this.events.get(DelegateVotedEvent).error(
+						context,
+						{
+							senderAddress,
+							delegateAddress: vote.delegateAddress,
+							amount: vote.amount,
+						},
+						PoSEventResult.VOTE_FAILED_INVALID_UNVOTE_PARAMETERS,
+					);
+
+					throw new Error('Invalid unvote: Cannot cast downvote to delegate who is not upvoted.');
 				}
 
-				voterData.sentVotes[originalUpvoteIndex].amount += vote.amount;
+				const aggregateVotedAmount = voterData.sentVotes[originalUpvoteIndex].amount + vote.amount;
+
+				if (aggregateVotedAmount < BigInt(0)) {
+					this.events.get(DelegateVotedEvent).error(
+						context,
+						{
+							senderAddress,
+							delegateAddress: vote.delegateAddress,
+							amount: vote.amount,
+						},
+						PoSEventResult.VOTE_FAILED_INVALID_UNVOTE_PARAMETERS,
+					);
 
-				if (voterData.sentVotes[originalUpvoteIndex].amount < BigInt(0)) {
-					throw new Error('The downvote amount cannot be greater than upvoted amount.');
+					throw new Error(
+						'Invalid unvote: The downvote amount cannot be greater than upvoted amount.',
+					);
 				}
 
 				// Delete entry when amount becomes 0
-				if (voterData.sentVotes[originalUpvoteIndex].amount === BigInt(0)) {
+				if (aggregateVotedAmount === BigInt(0)) {
```

Comment:
assignVote rewards should be called before this step, I think. The reason is, assignVoteRewards needs to access the vote object in the voter substore. Here, it is deleted, so the call to assignVoteRewards will consider wrong vote object.

Reply:
Line 264, makes sure to skip assignment of voteSharingCoefficients if the result of sum of `sentVote.amount` to `voter.sentVotes[votedDelegate].amount` results in zero.

Reply:
I think what @gkoumout means is `assignVoteRewards` should take the sharing coefficient before unvoting? so if we filter it here, we will change the `index` and lose the original voted value.
Order should be like below?
1. error check
2. assign reward
3. update sentVotes (amount + sharing coefficient)
4. create/sort/check pending unlocks

Reply:
also in this way, `aggregateVotedAmount` will not be needed

Reply:
@has5aan I am afraid there is some confusion. I agree on your comment on line 264. The problem is not assignments. The issue is, in line 224 you want to delete an entry in the voter substore if the *new* voted amount is 0. So then, when you call assignVoteRewards, the vote does not exist anymore...

In general, the point is that the order should be as @shuse2 mentioned IMO. The reason being:
 - The crucial is that step 2 (assign reward) should be before step 3 (update voter substore): the reason being that the assign rewards depends on the "old vote", so it should be first finished, before updating taking into account the new vote.
 - Step 4 (Unlocking) is recommended to be after step 3 (substores are updated), mostly for consistency reasons.

Reply:
@gkoumout & @shuse2 thank you! I overlooked that aspect, made the changes, to validate aggregate amount first and then `assignVoteRewards` and updating `sentVote.voteSharingCoefficient`, please confirm if the changes are correct. :pray:

Reply:
I am still not sure it is correct now (perhaps my .ts understanding is poor..). As far as I understand, in the current version in line 205 you update the voter substore for the vote amount and then in line 223 call assignVoteRewards. Is that correct?

If that's the case, the order is again wrong, since the update in voter substore is part of step 3 above and should happen *after* calling assingVoteRewards. The reason is that assignVote rewards will calculate rewards taking into account the current voted amount, which is not correct.

Reply:
ah good point, I misread that 🙇

Reply:
@gkoumout thanks :pray:. Please have another look.

Reply:
Looks good to me now!